libs/core/langchain_core/tracers/abstract_base_tracer.py

```diff
@@ -0,0 +1,562 @@
+"""Utilities for the root listener."""
+
+from __future__ import annotations
+
+import logging
+import sys
+import traceback
+from abc import ABC, abstractmethod
+from datetime import datetime, timezone
+from typing import (
+    TYPE_CHECKING,
+    Any,
+    Coroutine,
+    Dict,
+    List,
+    Literal,
+    Optional,
+    Sequence,
+    Set,
+    Tuple,
+    Union,
+    cast,
+)
+from uuid import UUID
+
+from tenacity import RetryCallState
+
+from langchain_core.exceptions import TracerException
+from langchain_core.load import dumpd
+from langchain_core.messages import BaseMessage
+from langchain_core.outputs import (
+    ChatGeneration,
+    ChatGenerationChunk,
+    GenerationChunk,
+    LLMResult,
+)
+from langchain_core.tracers.schemas import Run
+
+if TYPE_CHECKING:
+    from langchain_core.documents import Document
+
+logger = logging.getLogger(__name__)
+
+SCHEMA_FORMAT_TYPE = Literal["original", "streaming_events"]
+
+
+class AbstractBaseTracer(ABC):
```

Comment:
nit: base tracer is already abstract. I don't have a good suggestion here since the name is already occupied so probably fine.

Is this for internal use only if so, could we document it as private?

Reply:
Correct. Ideally it would be BaseTracer. 
Documenting as private, do you mean adding a note in the class description ?

Reply:
Changed to _TracerCore